### PR TITLE
[recipe, data] feat: TransferQueue - Support managing multiple data partitions for Train/Val/Test in controller

### PR DESCRIPTION
**Pull request by LLLLxmmm**
Support managing multiple data partitions for Train/Val/Test in controller

### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by 0oshowero0** [comment]
## Summary of changes
1. Data partition capability is available in [PR#98](https://github.com/TransferQueue/TransferQueue/pull/98). Now we can define logical data partitions to manage the train/val/test datasets.
2. We make storage backends pluggable in [PR#66](https://github.com/TransferQueue/TransferQueue/pull/66). You can try to integrate your own storage backend within TransferQueue interface.
3. `TransferQueueController` is greatly simplified. We disentangle the data retrieval logic from `TransferQueueController` [PR#101](https://github.com/TransferQueue/TransferQueue/pull/101). Now you can implement your own `Sampler` to control how to consume the data.
4. The initialization of TransferQueue is optimized.
5. TransferQueue can be installed through pypi: `pip install TransferQueue`
6. Add CI for TransferQueue.
7. Update docs.

## Accuracy comparison
image
image


## Previous PR
https://github.com/volcengine/verl/pull/3649

File: recipe/transfer_queue/ray_trainer.py [hunk 1]
```diff
@@ -1164,17 +1105,39 @@ def _stop_profiling(self, do_profile: bool) -> None:
...
+        global_seqlen_lst = data["attention_mask"].view(batch_size, -1).sum(-1)  # (train_batch_size,)
+        global_seqlen_lst = calculate_workload(global_seqlen_lst)
         world_size = self.actor_rollout_wg.world_size
-        global_partition_lst = get_seqlen_balanced_partitions(
-            global_seqlen_lst, k_partitions=world_size, equal_size=True
-        )
+        if keep_minibatch:
+            # Decouple the DP balancing and mini-batching.
+            minibatch_size = self.config.actor_rollout_ref.actor.get("ppo_mini_batch_size")
+            minibatch_num = len(global_seqlen_lst) // minibatch_size
```
[the remark this answers is not in the document]
Comment (0oshowero0) [reply recipe/transfer_queue/ray_trainer.py:1124]: solved

File: verl/experimental/agent_loop/agent_loop.py [hunk 2]
```diff
@@ -364,6 +364,11 @@ async def generate_sequences(self, batch: DataProto) -> DataProto:
         outputs = await asyncio.gather(*tasks)
 
         output = self._postprocess(outputs)
+
+        # Append Performance Data
+        output.meta_info["attention_mask_perf"] = output.batch["attention_mask"]
```
Comment (wuxibin89) [inline verl/experimental/agent_loop/agent_loop.py]: Please don't put torch.tensor into meta_info, it should be only contain small metrics.

Comment (0oshowero0) [reply verl/experimental/agent_loop/agent_loop.py]: Now maybe we need to temporarily comment some perf metrics in TQ recipe, since these perf metrics require the torch.tensor.

In the following PRs, we will pass TQ info throughout the `AgentLoop` so that when we need these data, we can use TQ client to retrieve them.

File: verl/utils/transferqueue_utils.py [hunk 3]
```diff
@@ -38,32 +38,24 @@ class BatchMeta:
...
 
 def get_transferqueue_client() -> "AsyncTransferQueueClient":
     return _TRANSFER_QUEUE_CLIENT
 
 
-def get_val_transferqueue_client() -> "AsyncTransferQueueClient":
-    return _VAL_TRANSFER_QUEUE_CLIENT
-
-
 def _run_async_in_temp_loop(async_func: Callable[..., Any], *args, **kwargs) -> Any:
```
Comment (wuxibin89) [inline verl/utils/transferqueue_utils.py:59]: TODO: we will make all actor async, so this can be cleanup later.
